File: ios/Classes/YandexBicycle.swift
import Flutter
import UIKit
import YandexMapsMobile

public class YandexBicycle: NSObject, FlutterPlugin {
  private let methodChannel: FlutterMethodChannel!
  private let pluginRegistrar: FlutterPluginRegistrar!
  private let bicycleRouter: YMKBicycleRouter!
  private var bicycleSessions: [Int: YandexBicycleSession] = [:]

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "yandex_mapkit/yandex_bicycle",
      binaryMessenger: registrar.messenger()
    )

    let plugin = YandexBicycle(channel: channel, registrar: registrar)

    registrar.addMethodCallDelegate(plugin, channel: channel)
  }

  public required init(channel: FlutterMethodChannel, registrar: FlutterPluginRegistrar) {
    self.pluginRegistrar = registrar
    self.methodChannel = channel
    self.bicycleRouter = YMKTransport.sharedInstance().createBicycleRouter()

    super.init()

    self.methodChannel.setMethodCallHandler(self.handle)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "requestRoutes":
      requestRoutes(call, result)
      break
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func requestRoutes(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    let params = call.arguments as! [String: Any]
    let sessionId = params["sessionId"] as! Int
    let requestPoints = (params["points"] as! [[String: Any]]).map {
      (pointParams) -> YMKRequestPoint in Utils.requestPointFromJson(pointParams)
    }
    let session = bicycleRouter.requestRoutes(
      with: requestPoints,
      type: YMKBicycleVehicleType(rawValue: (params["bicycleVehicleType"] as! NSNumber).uintValue)!,
      routeListener: {(bicycleResponse: [YMKBicycleRoute]?, error: Error?) -> Void in
        self.bicycleSessions[sessionId]?.handleResponse(bicycleResponse: bicycleResponse, error: error, result: result)
      }
    )

    bicycleSessions[sessionId] = YandexBicycleSession(
      id: sessionId,
      session: session,
      registrar: pluginRegistrar,
      onClose: { (id) in self.bicycleSessions.removeValue(forKey: id) }
    )
  }
}

File: android/src/main/java/com/unact/yandexmapkit/YandexBicycle.java
package com.unact.yandexmapkit;

import android.content.Context;

import androidx.annotation.NonNull;

import com.yandex.mapkit.RequestPoint;
import com.yandex.mapkit.transport.TransportFactory;
import com.yandex.mapkit.transport.bicycle.BicycleRouter;
import com.yandex.mapkit.transport.bicycle.Session;
import com.yandex.mapkit.transport.bicycle.VehicleType;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

public class YandexBicycle implements MethodCallHandler {
    private final BicycleRouter bicycleRouter;
    private final BinaryMessenger binaryMessenger;
    @SuppressWarnings({"MismatchedQueryAndUpdateOfCollection"})
    private final Map<Integer, YandexBicycleSession> bicycleSessions = new HashMap<>();

    public YandexBicycle(Context context, BinaryMessenger messenger) {
        TransportFactory.initialize(context);

        bicycleRouter = TransportFactory.getInstance().createBicycleRouter();
        binaryMessenger = messenger;
    }

    @Override
    @SuppressWarnings({"SwitchStatementWithTooFewBranches"})
    public void onMethodCall(MethodCall call, @NonNull Result result) {
        switch (call.method) {
            case "requestRoutes":
                requestRoutes(call, result);
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    @SuppressWarnings({"unchecked", "ConstantConditions"})
    private void requestRoutes(final MethodCall call, final Result result) {
        Map<String, Object> params = (Map<String, Object>) call.arguments;
        Integer sessionId = (Integer) params.get("sessionId");
        List<RequestPoint> points = new ArrayList<>();
        for (Map<String, Object> pointParams : (List<Map<String, Object>>) params.get("points")) {
            points.add(Utils.requestPointFromJson(pointParams));
        }

        Session session = bicycleRouter.requestRoutes(
                points,
                VehicleType.values()[(Integer) params.get("bicycleVehicleType")],
                new YandexBicycleListener(result)
        );

        YandexBicycleSession bicycleSession = new YandexBicycleSession(
                sessionId,
                session,
                binaryMessenger,
                new BicycleCloseListener()
        );

        bicycleSessions.put(sessionId, bicycleSession);
    }

    public class BicycleCloseListener {
        public void onClose(int id) {
            bicycleSessions.remove(id);
        }
    }
}
